### PR TITLE
Implement calculate_roots

### PR DESCRIPTION
**Pull request by Davidson-Souza**
This PR implements a very similar function to what we have in pytreexo, where we delete without using `get_proof_positions`. In this implementation, I basically tweaked the existing `create_root_candidates` to the new semantics, and replaced it everywhere. All tests are passing, I also added another two tests for cases where the null hashes are heavily propagated up.

### CODE REVIEW

File: src/accumulator/proof.rs [hunk 1]
```diff
@@ -305,21 +330,27 @@ impl Proof {
...
+                        };
 
-                    let next_to_prove = util::parent(pos, total_rows);
+                        let next_to_prove = util::parent(pos, total_rows);
 
-                    Proof::sorted_push(&mut nodes, (next_to_prove, hash));
+                        Proof::sorted_push(&mut nodes, (next_to_prove, hash));
+                        continue;
+                    } else {
+                        // If none of the above, push a nil hash upwards
```
Comment (kcalvinalvin) [inline src/accumulator/proof.rs:345]: Technically not a nil right? Maybe just "push a default hash upwards"

File: src/accumulator/proof.rs [hunk 2]
```diff
@@ -289,10 +301,23 @@ impl Proof {
...
-                        let next_to_prove = util::parent(pos, total_rows);
-
-                        Proof::sorted_push(&mut nodes, (next_to_prove, hash));
+                        // There are three possible cases: the current hash is null,
+                        // and the sibling is present, we push the sibling to targets.
+                        // If The sibling is null, we push the current node.
+                        // If none of them is null, we compute the parent hash of both siblings
+                        // and push this to the next target.
+                        if hash == sha256::Hash::default() {
+                            let next_to_prove = util::parent(pos, total_rows);
```
Comment (kcalvinalvin) [inline src/accumulator/proof.rs:310]: `next_to_prove` is always going to be `util::parent(pos, total_rows)` no matter what the hash is. And it's just a u64 so it's on the stack too.

Just declaring it one right after the `is_root_position` would be better in my opinion.

**Review by kcalvinalvin — CHANGES_REQUESTED** [review]
Overall looks good. Have some few suggestions

File: src/accumulator/proof.rs [hunk 3]
```diff
@@ -245,7 +251,13 @@ impl Proof {
             },
         ))
     }
-    pub fn create_root_candidates(
+
+    /// This function computes a set of roots from a proof.
+    /// If some target's hashes are null, then it computes the roots after
+    /// those targets are deleted.
+    /// It's the caller's responsibility to null out the targets if desired by
```
[the remark this answers is not in the document]
Comment (Davidson-Souza) [reply src/accumulator/proof.rs:258]: I actually think this function should not be public, not sure why the API consumers would want such a lol-level function. What do you think?

File: src/accumulator/proof.rs (hunk 1 above)
Comment (Davidson-Souza) [reply src/accumulator/proof.rs:345]: Yeah, it's the default hash. You can't have a nil in safe rust outside an option.

**Comment by Davidson-Souza** [comment]
@kcalvinalvin do you think it's worth keeping this PR and then making a new with `calculate_hashes` you just mentioned Monday, or add the whole new semantics here?

**Comment by kcalvinalvin** [comment]
> @kcalvinalvin do you think it's worth keeping this PR and then making a new with `calculate_hashes` you just mentioned Monday, or add the whole new semantics here?

I think keeping this PR and refactoring in a follow up makes sense since we need this change anyways. Also, I don't think I pushed that code yet. Will do that soon.

But getting rid of `proof_after_deletion` in this PR (probably in a separate commit) makes sense since it won't be used anymore.

File: src/accumulator/proof.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (kcalvinalvin) [reply src/accumulator/proof.rs:258]: Yeah it'd be better if it wasn't public. My assumption was that it's public because of rust modules stuff. Would prefer it to be private.

[the remark this answers is not in the document]
Comment (Davidson-Souza) [reply src/accumulator/proof.rs:149]: I made it pub(crate), so it can only be used inside the crate and not by API consumers. We need `calculate_roots` for stump updates, and thus it can't be private.

**Comment by Davidson-Souza** [comment]
Rebased on top of main, and addressed the comments.

File: src/accumulator/proof.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (kcalvinalvin) [reply src/accumulator/proof.rs:149]: Ah ok. That sounds ok.

**Comment by kcalvinalvin** [comment]
ACK 8e5f7974e967c587854cb186290f95a4be0ea675

**Comment by kcalvinalvin** [comment]
> @kcalvinalvin do you think it's worth keeping this PR and then making a new with `calculate_hashes` you just mentioned Monday, or add the whole new semantics here?

It's not merged into master yet but here's the code for `calculate_hashes`.
https://github.com/utreexo/utreexo/pull/58
